crates/turborepo-lockfiles/src/npm.rs

```diff
@@ -0,0 +1,365 @@
+            // the package
+            format!("{}/node_modules/{}", workspace_path, name),
+            // Next we check for a top level version of the package
+            format!("node_modules/{}", name),
+        ];
+        possible_keys
+            .into_iter()
+            .filter_map(|key| {
+                self.packages.get(&key).map(|pkg| {
+                    let version = pkg.version.clone().unwrap_or_default();
+                    Ok(Package { key, version })
+                })
+            })
+            .next()
+            .transpose()
+    }
+
+    fn all_dependencies(&self, key: &str) -> Result<Option<HashMap<String, &str>>, Error> {
+        self.packages
+            .get(key)
+            .map(|pkg| {
+                pkg.dep_keys()
+                    .filter_map(|name| {
+                        Self::possible_npm_deps(key, name)
+                            .into_iter()
+                            .find_map(|possible_key| {
+                                self.packages.get(&possible_key).map(|entry| {
+                                    let version = entry.version.as_deref().ok_or_else(|| {
+                                        Error::MissingVersion("".into(), possible_key.clone())
+                                    })?;
+                                    Ok((possible_key, version))
+                                })
+                            })
+                    })
+                    .collect()
+            })
+            .transpose()
+    }
+}
+
+impl NpmLockfile {
+    pub fn load(content: &[u8]) -> Result<Self, Error> {
+        let lockfile: NpmLockfile = serde_json::from_slice(content)?;
+
+        // We don't support lockfiles without 'packages' as older versions
+        // required reading through the contents of node_modules in order
+        // to resolve dependencies.
+        // See https://github.com/npm/cli/blob/9609e9eed87c735f0319ac0af265f4d406cbf800/workspaces/arborist/lib/shrinkwrap.js#L674
+        if lockfile.lockfile_version <= 1
+            || (lockfile.packages.is_empty() && !lockfile.dependencies.is_empty())
+        {
+            Err(Error::UnsupportedNpmVersion)
+        } else {
+            Ok(lockfile)
+        }
+    }
+
+    fn get_package(&self, package: impl AsRef<str>) -> Result<&NpmPackage, Error> {
+        let pkg_str = package.as_ref();
+        self.packages
+            .get(pkg_str)
+            .ok_or_else(|| Error::MissingPackage(pkg_str.to_string()))
+    }
+
+    pub fn subgraph(
+        &self,
+        workspace_packages: &[String],
+        packages: &[String],
+    ) -> Result<Self, Error> {
+        let mut pruned_packages = HashMap::with_capacity(packages.len());
+        for pkg_key in packages {
+            let pkg = self.get_package(pkg_key)?;
+            pruned_packages.insert(pkg_key.to_string(), pkg.clone());
+        }
+        if let Some(root) = self.packages.get("") {
```

Comment:
Does is make sense to encode the root into the type system? I'm pretty sure with pointer types this has no perf overhead as rust just encodes the Root case to be the null value, but it is of course a bit more complicated.

```rust
enum Package<'a> {
  Root,
  Named(&'a str),
}
```

Reply:
That's getting a little too clever for my taste as I want `NpmLockfile` to be a transparent in-memory version of the `package-lock.json` on disk and `package-lock.json` has a `""` entry in the `packages` section

Reply:
> package-lock.json has a "" entry in the packages section

Didn't know that, I assumed that that was a turbo convention, TY :)